enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The Assisted Installer currently runs as a SaaS on cloud.redhat.com, enabling
+users to deploy OpenShift clusters with certain customizations, particularly on
+bare metal hardware. It is necessary to bring those capabilities on-premise in
+users' clusters for two purposes:
+
+1. A cluster created by the Assisted Installer needs the ability to add workers
+on day 2.
+2. Multi-cluster management, such as through hive and ACM, should be
+able to utilize the capabilities of the Assisted Installer.
+
+This enhancement proposes the first iteration of running the Assisted Installer
+in end-user clusters to meet the purposes above.
+
+## Motivation
+
+### Goals
+
+* Expose the Assisted Installer's capabilities as a kubernetes-native API.
+* Enable any OpenShift cluster deployed by the Assisted Installer to add worker nodes on day 2.
+* Enable multi-cluster management tooling to create new clusters using Assisted Installer.
+* Utilize [metal3](https://metal3.io/) to boot the assisted discovery ISO for bare-metal deployments.
+* Ensure the design is be extensible for installing on other platforms.
+* Utilize the Assisted Service and Agent with minimal changes to avoid
+adversely affecting the cloud.redhat.com deployment.  While making a version
+that is Kubernetes-native with no SQL DB is on the table for the long term,
+it is not feasible to implement and maintain two versions in the near-term.
+
+### Non-Goals
+
+* Solve central machine management. That can be done with this effort, or after
+this effort, but it is not strictly a requirement in order to deliver the
+goals stated above.
+* Run metal3 components on a non-baremetal cluster.
+
+## Proposal
+
+### User Stories
+
+#### Day 2 Add Worker
+
+After deploying a cluster with Assisted Installer, I can add workers from bare
+metal hardware on day two using a similar workflow and tool set. I can either
+obtain the discovery live ISO and use my own methods to boot it, or I can use
+the baremetal-operator to boot the live ISO automatically.
+
+#### Multi-cluster
+
+As a user of Red Hat's multi-cluster mangement tools, I can use the assisted
+installation agent-based workflow to create clusters from a pool of bare metal
+inventory.
+
+### Implementation Details/Notes/Constraints
+
+#### Assisted Installer k8s-native API
+
+The Assisted Installer has a [REST
+API](https://generator.swagger.io/?url=https://raw.githubusercontent.com/openshift/assisted-service/master/swagger.yaml)
+that is available at cloud.redhat.com. The service is implemented with a
+traditional relational database. In order to integrate with OpenShift
```

Comment:
Hmm.. I don't follow The reasoning here that there is a requirement to be kube API. Things like hive can interact with some rest API like it does today with cloud provider APIs. Maybe that will also allow just moving the entire stack as- is.

Reply:
Another thought against CRD;; since the agents & hive will have to talk to these APIs, in the the CRD approach it means the discovery is tied to the cluster running the setup, ie.. the apiserver. If we use generic rest that would allow the discovery to be different from cluster address to only some address, which is more locality independent.

Reply:
As also answered above, hive won't be the only user of this API.

> If we use generic rest that would allow the discovery to be different from cluster address to only some address, which is more locality independent.

Could you elaborate on that? I'm not sure what the advantage is that you're describing.

Reply:
I think the point there is that Hive and the Assisted Installer service would not need to run in the same cluster together if Hive was using the existing REST API.

Reply:
I see. I'm not aware of a desire or an advantage to run hive and the assisted installer service in separate clusters. If anything, the openshift architecture trend is to run more and more management functions together in a central hub cluster.

Reply:
The assisted installer API is starting to feel a bit like a "cloud" API, since there are compute resources and an API to provision them so they join a cluster. So, I can see the perspective of wanting to think about using it that way. If it's not a part of the planned architecture, you could list it as an alternative at the end of the doc so we remember why it's not being considered.

Reply:
I added an alternative called "Expose a non-k8s-native REST API" explaining that the CRD-based API fits openshift's approach to management APIs. If there is still some thinking that this could all work with just the REST API and not with any CRDs, it would help to see a more detailed vision for how that would work.